Acuan Modul: https://taudata.blogspot.com/2022/10/evda-04-data-wrangling.html
             https://colab.research.google.com/github/taudataid/eLearning/blob/master/E(V)DA%2004%20-%20Data%20Wrangling.ipynb


# **Import Data**

In [ ]:
import warnings; warnings.simplefilter('ignore')
import pandas as pd

In [ ]:
try: # Running Locally, yakinkan "file_" berada di folder "data"
    mengajar = pd.read_csv('data/mengajar.csv', encoding='utf8')
    dosen = pd.read_csv('data/tb_dosen.csv', encoding='utf8')
    kelas = pd.read_csv('data/tb_kelas.csv', encoding='utf8')
    krs = pd.read_csv('data/tb_krs.csv', encoding='utf8')
    mahasiswa = pd.read_csv('data/tb_mahasiswa.csv', encoding='utf8')
    matkul = pd.read_csv('data/tb_mata_kuliah.csv', encoding='utf8')
    nilai = pd.read_csv('data/tb_nilai.csv', encoding='utf8')
except: # Running in Google Colab
    !mkdir data
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_nilai.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/mengajar.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_dosen.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_kelas.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_krs.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_mahasiswa.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_mata_kuliah.csv
    mengajar = pd.read_csv('data/mengajar.csv', encoding='utf8')
    dosen = pd.read_csv('data/tb_dosen.csv', encoding='utf8')
    kelas = pd.read_csv('data/tb_kelas.csv', encoding='utf8')
    krs = pd.read_csv('data/tb_krs.csv', encoding='utf8')
    mahasiswa = pd.read_csv('data/tb_mahasiswa.csv', encoding='utf8')
    matkul = pd.read_csv('data/tb_mata_kuliah.csv', encoding='utf8')
    nilai = pd.read_csv('data/tb_nilai.csv', encoding='utf8')

mengajar.shape, dosen.shape, kelas.shape, krs.shape, mahasiswa.shape, matkul.shape, nilai.shape

((14, 4), (4, 6), (4, 4), (7, 6), (15, 8), (9, 3), (33, 7))

# **Table Information**

## **Table mengajar**

In [ ]:
mengajar

NIDN kode_MK Semester  Tahun
0    123   KAL01   ganjil   2020
1    123   KAL01   ganjil   2019
2    123   STA01    genap   2020
3    123   MAT01    genap   2020
4    123   STA02   ganjil   2019
5    234    OR01   ganjil   2019
6    234    CS01    genap   2020
7    787   KAL01   ganjil   2021
8    787   KAL01   ganjil   2022
9    787   STA02   ganjil   2021
10   345   MAT73    genap   2020
11   345   MAT53    genap   2019
12   345   STA02   ganjil   2019
13   345   KAL02   ganjil   2020

In [ ]:
mengajar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   NIDN      14 non-null     int64 
 1   kode_MK   14 non-null     object
 2   Semester  14 non-null     object
 3   Tahun     14 non-null     int64 
dtypes: int64(2), object(2)
memory usage: 576.0+ bytes


## **Table dosen**

In [ ]:
dosen

NIDN           nama             email Pendidikan_Terakhir  Telephone  \
0   123   Andri Basuki   anbr@googol.com                  S3   11230000   
1   234       zulfikar   dosen@email.com                  S3   22340000   
2   787           yudi    yudi@email.com                  S2   55670000   
3   345  Annisa Fitria  annniza@yyyy.com                  S2   88970000   

  Jenis_Kelamin  
0          Pria  
1          Pria  
2          Pria  
3        Wanita

In [ ]:
dosen.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   NIDN                 4 non-null      int64 
 1   nama                 4 non-null      object
 2   email                4 non-null      object
 3   Pendidikan_Terakhir  4 non-null      object
 4   Telephone            4 non-null      int64 
 5   Jenis_Kelamin        4 non-null      object
dtypes: int64(2), object(4)
memory usage: 320.0+ bytes


## **Table kelas**

In [ ]:
kelas

kode_kelas nama_kelas lokasi_kelas  Kapasitas
0      SIS1J  Reguler A     Gedung D         10
1       D101  Reguler A     Gedung D         15
2       D212  Reguler B     Gedung D         40
3       P601  Reguler B   Gedung PLT         30

In [ ]:
kelas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   kode_kelas    4 non-null      object
 1   nama_kelas    4 non-null      object
 2   lokasi_kelas  4 non-null      object
 3   Kapasitas     4 non-null      int64 
dtypes: int64(1), object(3)
memory usage: 256.0+ bytes


## **Table krs**

In [ ]:
krs

id id_mata_kuliah kode_kelas     nim  tahun semester
0   1          KAL01      SIS1J  193001   2021    genap
1   2          STA01       D101  193001   2021    genap
2   3           OR01       D212  193001   2021    genap
3   4          MK002       P601  194001   2021    genap
4   5          MK002       P601  194001   2021    genap
5   6          MK003      SIS1J  194001   2021    genap
6   7          MK004       D101  194001   2021    genap

In [ ]:
krs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   id              7 non-null      int64 
 1   id_mata_kuliah  7 non-null      object
 2   kode_kelas      7 non-null      object
 3   nim             7 non-null      int64 
 4   tahun           7 non-null      int64 
 5   semester        7 non-null      object
dtypes: int64(3), object(3)
memory usage: 464.0+ bytes


## **Table mahasiswa**

In [ ]:
mahasiswa

nim            nama                    email  Gender Asal_Sekolah  \
0   194015   Muhammad Amir       Samarinda@inet.com    Pria          SMU   
1   194001   Samsul Kurnia        Bontang@email.com    Pria          SMU   
2   194002  Bambang Arifin         Sangata@inet.com    Pria          SMU   
3   194003            Udin            Udin@inet.com    Pria           MA   
4   194004            Iwan            Iwan@inet.com    Pria           MA   
5   194005  Bambang Arifin  Bambang Arifin@inet.com    Pria          SMK   
6   194006            Wati            Wati@inet.com  Wanita          SMK   
7   194007             Ani             Ani@inet.com  Wanita          SMK   
8   194008       Rahmawati       Rahmawati@inet.com  Wanita          SMU   
9   194009            Asep            Asep@inet.com    Pria          SMU   
10  194010            Rina            rina@inet.com  Wanita           MA   
11  194011            Sari            sari@inet.com  Wanita           MA   
12  194012            Bowo            Bowo@inet.com    Pria          SMU   
13  194013            Budi            Budi@inet.com    Pria          SMK   
14  194014           Putri           putri@inet.com  Wanita          SMK   

    Telephone  Angkatan  Asal_Daerah  
0    12345678      2019  Luar Daerah  
1    12345679      2019        Lokal  
2    12345680      2019  Luar Daerah  
3    12345681      2019        Lokal  
4    12345682      2019  Luar Daerah  
5    12345683      2019        Lokal  
6    12345684      2019        Lokal  
7    12345685      2019  Luar Daerah  
8    12345686      2020        Lokal  
9    12345687      2020  Luar Daerah  
10   12345688      2020        Lokal  
11   12345689      2020        Lokal  
12   12345690      2020  Luar Daerah  
13   12345691      2020        Lokal  
14   12345692      2020        Lokal

In [ ]:
mahasiswa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   nim           15 non-null     int64 
 1   nama          15 non-null     object
 2   email         15 non-null     object
 3   Gender        15 non-null     object
 4   Asal_Sekolah  15 non-null     object
 5   Telephone     15 non-null     int64 
 6   Angkatan      15 non-null     int64 
 7   Asal_Daerah   15 non-null     object
dtypes: int64(3), object(5)
memory usage: 1.1+ KB


## **Table matkul**

In [ ]:
matkul

id_mata_kuliah           nama_mata_kuliah  SKS
0          KAL01             Kalkulus Dasar    3
1          STA01       Statistika Elementer    3
2          MAT01           Matematika Dasar    3
3          STA02      Statistika Matematika    4
4           OR01         Operation Research    4
5           CS01  Pendahuluan Ilmu Komputer    3
6          MAT73            Aljabar Abstrak    4
7          MAT53             Aljabar linear    3
8          KAL02            Kalkulus Lanjut    4

In [ ]:
matkul.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   id_mata_kuliah    9 non-null      object
 1   nama_mata_kuliah  9 non-null      object
 2   SKS               9 non-null      int64 
dtypes: int64(1), object(2)
memory usage: 344.0+ bytes


## **Table nilai**

In [ ]:
nilai

nim id_mata_kuliah  NIDN  nilai Semester  Tahun     rand
0   194015          KAL01   123     70   ganjil   2019  0.50463
1   194001          KAL01   123     80   ganjil   2019      NaN
2   194002          KAL01   123     85   ganjil   2019      NaN
3   194003          KAL01   123     90   ganjil   2019      NaN
4   194004          KAL01   123     92   ganjil   2019      NaN
5   194005          KAL01   123     91   ganjil   2019      NaN
6   194006          KAL01   123     93   ganjil   2019      NaN
7   194007          KAL01   123     85   ganjil   2019      NaN
8   194008          KAL01   787     60   ganjil   2020      NaN
9   194009          KAL01   787     63   ganjil   2020      NaN
10  194010          KAL01   787     62   ganjil   2020      NaN
11  194011          KAL01   787     61   ganjil   2020      NaN
12  194012          KAL01   787     70   ganjil   2020      NaN
13  194013          KAL01   787     90   ganjil   2020      NaN
14  194014          KAL01   787     88   ganjil   2020      NaN
15  194004           CS01   234     90    genap   2020      NaN
16  194005           CS01   234     92    genap   2020      NaN
17  194006           CS01   234     85    genap   2020      NaN
18  194007           CS01   234     55    genap   2020      NaN
19  194008           CS01   234     60    genap   2020      NaN
20  194009           CS01   234     91    genap   2020      NaN
21  194010           CS01   234     90    genap   2020      NaN
22  194015          STA02   345     40   ganjil   2021      NaN
23  194002          STA02   345     50   ganjil   2021      NaN
24  194004          STA02   345     56   ganjil   2021      NaN
25  194007          STA02   345     70   ganjil   2021      NaN
26  194008          STA02   345     72   ganjil   2021      NaN
27  194011          STA02   345     31   ganjil   2021      NaN
28  194014          STA02   345     88   ganjil   2021      NaN
29  194001          KAL02   345     55   ganjil   2020      NaN
30  194002          KAL02   345     60   ganjil   2020      NaN
31  194004          KAL02   345     65   ganjil   2020      NaN
32  194007          KAL02   345     74   ganjil   2020      NaN

In [ ]:
nilai.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33 entries, 0 to 32
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   nim             33 non-null     int64  
 1   id_mata_kuliah  33 non-null     object 
 2   NIDN            33 non-null     int64  
 3   nilai           33 non-null     int64  
 4   Semester        33 non-null     object 
 5   Tahun           33 non-null     int64  
 6   rand            1 non-null      float64
dtypes: float64(1), int64(4), object(2)
memory usage: 1.9+ KB


# **Task 01:**

Apakah ada dosen yang cenderung memberikan nilai cenderung lebih tinggi (atau lebih rendah) dibandingkan dosen lainnya?

In [ ]:
# Code Task 01

# membuat table nilai rata-rata masing-masing kolom table nilai
nilai_dosen = nilai.groupby(nilai['NIDN']).mean()[['nilai']]
nilai_dosen = nilai_dosen.rename({'nilai': 'mean_nilai'}, axis=1)

# merge left table dosen dan nilai_dosen dengan key: NIDN
nilai_dosen = pd.merge(dosen, nilai_dosen, left_on='NIDN', right_index=True)
nilai_dosen = nilai_dosen[['NIDN', 'nama', 'mean_nilai']]

# membuat table nilai standar deviasi masing-masing kolom table nilai
sb = nilai.groupby(nilai['NIDN']).std()[['nilai']]
sb = sb.rename({'nilai': 'std_nilai'}, axis=1)

# concat nilai_dosen dan sb
nilai_dosen = nilai_dosen.set_index(['NIDN'])
nilai_dosen = pd.concat([nilai_dosen, sb], axis=1)
nilai_dosen

nama  mean_nilai  std_nilai
NIDN                                      
123    Andri Basuki   85.750000   7.741355
234        zulfikar   80.428571  15.883504
787            yudi   70.571429  13.010984
345   Annisa Fitria   60.090909  16.220077

## **Interpretasi 01:**

1. Dalam data yang disajikan, terlihat perbedaan dalam rata-rata penilaian yang diberikan oleh setiap dosen. **Andri Basuki** memperoleh rata-rata tertinggi, yaitu 85.750000. Di posisi kedua adalah **Zulfikar** dengan rata-rata penilaian sebesar 80.428571. Posisi ketiga ditempati oleh **Yudi** dengan rata-rata penilaian sebesar 70.571429. Sedangkan di posisi keempat adalah **Annisa Fitria** dengan rata-rata penilaian sebesar 60.090909.

2. Rata-rata penilaian yang diberikan oleh **Andri Basuki** memiliki standar deviasi terkecil (7.741355), menunjukkan bahwa nilai-nilai tersebut lebih mendekati rata-rata dan variasi data tidak begitu tersebar. Dalam kata lain, dosen tersebut cenderung memberikan nilai secara lebih merata (lebih seragam).

3. Sementara rata-rata penilaian yang diberikan oleh **Annisa Fitria** memiliki standar deviasi terbesar (16.220077), menandakan bahwa variasi data akan lebih tersebar. Dalam kata lain, dosen tersebut cenderung memberikan nilai secara lebih objektif sesuai kemampuan mahasiswa.

Kesimpulan:

Dengan demikian, dapat disimpulkan bahwa terdapat perbedaan dalam cara pemberian penilaian antara satu dosen dengan dosen lainnya. Beberapa dosen cenderung memberikan penilaian yang relatif lebih tinggi atau lebih rendah dibandingkan dengan dosen lainnya.

# **Task 02:**



In [ ]:
# Code Task 02

# merge table dosen dan nilai dengan key: NIDN
merged_data = pd.merge(dosen, nilai, on='NIDN')
result_data = merged_data[['Pendidikan_Terakhir', 'nilai']]

# membuat table nilai rata-rata masing-masing kolom table result_data
mean_data = result_data.groupby(merged_data['Pendidikan_Terakhir']).mean()[['nilai']]
mean_data = mean_data.rename({'nilai': 'mean_nilai'}, axis=1)

# membuat table nilai standar deviasi masing-masing kolom table merge_data
std_data = merged_data.groupby(merged_data['Pendidikan_Terakhir']).std()[['nilai']]
std_data = std_data.rename({'nilai': 'std_nilai'}, axis=1)

# concat table mean_data dan std_data
final_data = pd.concat([mean_data, std_data], axis=1)
final_data

mean_nilai  std_nilai
Pendidikan_Terakhir                       
S2                    64.166667  15.561075
S3                    83.266667  12.068061

##Interpretasi 02:

1. Dari data yang dianalisis, terlihat bahwa** rata-rata penilaian dosen yang memiliki gelar pendidikan terakhir S3 lebih tinggi daripada yang memiliki gelar pendidikan terakhir S2**. Rata-rata penilaian dosen dengan gelar S3 adalah 83.266667, sementara rata-rata penilaian dosen dengan gelar S2 adalah 64.166667. Perubahan nilai standar deviasi terhadap nilai rata-rata adalah semakin rendah nilai standar deviasi, semakin mendekati nilai rata-rata (dengan penyebaran yang lebih terkonsentrasi). Sebaliknya, jika nilai standar deviasi tinggi, rentang variasi data akan semakin luas (menyebar secara signifikan).

2. Rata-rata penilaian dosen dengan gelar **S3** cenderung lebih dekat dengan nilai rata-rata, ditandai dengan nilai **standar deviasi yang rendah** (12).Sementara itu, **rata-rata penilaian dosen dengan gelar S2 memiliki variasi data yang lebih luas** dengan nilai standar deviasi yang tinggi (15). Karena dilihat dari nilai standar deviasi rata-rata penilaian dosen dengan gelar S3 dan S2 yang tidak terlalu jauh, maka kemungkinan nilai tersebut diberikan secara objektif.

3. **Nilai standar deviasi yang besar mencerminkan sebaran data yang luas**. Semakin besar nilai standar deviasinya, semakin bervariasi nilai-nilainya atau semakin tidak akurat nilai rata-ratanya.

**Kesimpulan**:

Secara keseluruhan, dapat disimpulkan bahwa terdapat perbedaan yang signifikan dalam pemberian penilaian antara dosen-dosen yang memiliki gelar pendidikan terakhir S3 dan S2. Namun, persebaran dari nilai tersebut hampir sama sehingga nilai tersebut diberikan secara objektif.

# **Task 03:**



In [ ]:
# Code Task 03

# buat table baru yang komponennya nilai kolom nim dan nilai dari table nilai
nilai_extracted = nilai[['nim','nilai']]

# buat table baru yang komponennya nilai kolom nim dan asal sekolah dari table mahasiswa
mahasiswa_extracted = mahasiswa[['nim', 'Asal_Sekolah']]

# merge left table nilai_extracted dan mahasiswa_extracted dengan key: nim
merged_data = pd.merge(mahasiswa_extracted, nilai_extracted, on='nim', how='left')

# membuat table nilai rata-rata masing-masing kolom table merged_data
mean_data = merged_data.groupby('Asal_Sekolah')[['nilai']].agg('mean')
mean_data = mean_data.rename({'nilai': 'mean_nilai'}, axis=1)

# membuat table nilai standar deviasi masing-masing kolom table merged_data
std_data = merged_data.groupby(merged_data['Asal_Sekolah']).std()[['nilai']]
std_data = std_data.rename({'nilai': 'std_nilai'}, axis=1)

# concat table mean_data dan std_data
final_data = pd.concat([mean_data, std_data], axis=1)
final_data

mean_nilai  std_nilai
Asal_Sekolah                       
MA             70.777778  21.111477
SMK            82.818182  11.771306
SMU            65.846154  14.200036

## **Interpretasi 03:**

1.   Dari data tabel, terlihat bahwa mahasiswa yang berasal dari latar belakang pendidikan **SMK memiliki rata-rata nilai tertinggi** selama masa perkuliahan, yakni 82.81. Pada peringkat kedua tertinggi, terdapat mahasiswa yang pendidikan terakhirnya berasal dari MA (Madrasah), sementara peringkat terakhir ditempati oleh mahasiswa yang latar belakang pendidikan terakhirnya adalah SMU, dengan nilai sebesar 65.84.

2.   Rata-rata nilai dari mahasiswa yang berasal dari **SMK memiliki standar deviasi terendah** (11.77), menandakan bahwa nilai-nilai tersebut lebih mendekati rata-rata dan variasi datanya tidak terlalu tersebar.

3. Di sisi lain, rata-rata nilai dari mahasiswa yang berasal dari **MA memiliki standar deviasi tertinggi** (21.11), yang menunjukkan bahwa variasi data akan semakin tersebar.


**Kesimpulan:**

Dengan demikian, dapat disimpulkan bahwa performa akademik, yang dinilai berdasarkan rata-rata nilai selama perkuliahan, memiliki perbedaan yang cukup signifikan tergantung pada latar belakang pendidikan terakhir dari para mahasiswa.

# **Task 04:**



In [ ]:
# Code Task 04

# buat table baru yang komponennya nilai kolom nim dan nilai dari table nilai
nilai_extracted = nilai[['nim','nilai']]

# buat table baru yang komponennya nilai kolom nim dan gender dari table mahasiswa
mahasiswa_extracted = mahasiswa[['nim', 'Gender']]

# merge left table nilai_extracted dan mahasiswa_extracted dengan key: nim
merged_data = pd.merge(mahasiswa_extracted, nilai_extracted, on='nim', how='left')

# membuat table nilai rata-rata masing-masing kolom table merged_data
mean_data = merged_data.groupby('Gender')[['nilai']].agg('mean')
mean_data = mean_data.rename({'nilai': 'mean_nilai'}, axis=1)

# membuat table nilai standar deviasi masing-masing kolom table merged_data
std_data = merged_data.groupby(merged_data['Gender']).std()[['nilai']]
std_data = std_data.rename({'nilai': 'std_nilai'}, axis=1)

# concat table mean_data dan std_data
final_data = pd.concat([mean_data, std_data], axis=1)
final_data

mean_nilai  std_nilai
Gender                       
Pria     73.888889  17.146047
Wanita   71.600000  17.112235

## **Interpretasi 04**

1. Berdasarkan nilai rata-rata yang diperoleh, yakni untuk kelompok Pria (mahasiswa) dan Wanita (mahasiswi), dalam tabel terlihat bahwa** nilai rata-rata mahasiswa pria lebih tinggi** (73.89) daripada nilai rata-rata mahasiswi (71.60).

2. Rata-rata nilai dari **mahasiswa memiliki standar deviasi kecil** (17.14), yang menunjukkan bahwa nilai-nilai tersebut lebih mendekati rata-rata dan variasi datanya tidak terlalu tersebar.

3. Sementara rata-rata nilai dari mahasiswi juga memiliki standar deviasi kecil (17.11), menandakan bahwa variasi data lebih mendekati rata-rata dan variasinya tidak terlalu tersebar.

Kesimpulan:

Secara keseluruhan, dapat disimpulkan bahwa tidak ada cukup bukti yang menyatakan bahwa terdapat perbedaan yang signifikan antara nilai mahasiswa dan nilai mahasiswi.

# **Task 05:**



In [ ]:
# Code Task 05

# merge table nilai dan mahasiswa dengan key: nim
merged_df = pd.merge(nilai, mahasiswa, on='nim')\

# buat table baru yang komponennya nilai kolom angakatn, id mata kuliah, dan nilai dari merged_df
merged_df = merged_df[['Angkatan', 'id_mata_kuliah', 'nilai']]

# membuat table nilai rata-rata masing-masing kolom table merged_df
mean_data = merged_df.groupby(['Angkatan', 'id_mata_kuliah'])[['nilai']].mean()
mean_data = mean_data.rename({'nilai': 'mean_nilai'}, axis=1)

# membuat table nilai nilai variansi masing-masing kolom table merged_df
std_data = merged_df.groupby(['Angkatan', 'id_mata_kuliah']).std()[['nilai']]
std_data = std_data.rename({'nilai': 'std_nilai'}, axis=1)

# concat table mean_data dan std_data
final_data = pd.concat([mean_data, std_data], axis=1)
final_data


mean_nilai  std_nilai
Angkatan id_mata_kuliah                       
2019     CS01             80.500000  17.253019
         KAL01            85.750000   7.741355
         KAL02            63.500000   8.103497
         STA02            54.000000  12.543258
2020     CS01             80.333333  17.616280
         KAL01            70.571429  13.010984
         STA02            63.666667  29.399546

In [ ]:
# buat table rata-rata dan standar deviasi masing-masing angkatan
a = pd.merge(mahasiswa[['nama','nim','Angkatan']].set_index('nim'), nilai[nilai['id_mata_kuliah'] != 'KAL02'][['nim','nilai']], right_on='nim', left_index=True).groupby(['Angkatan'])['nilai'].mean()
sb = pd.merge(mahasiswa[['nama','nim','Angkatan']].set_index('nim'), nilai[['nim','nilai']], right_on='nim', left_index=True).groupby(['Angkatan'])['nilai'].std().rename('std')
pd.merge(a,sb.reset_index(), right_on='Angkatan', left_index=True)[['Angkatan','nilai','std']]

Angkatan      nilai        std
0      2019  76.500000  16.695099
1      2020  71.230769  17.767732

##Interpretasi 05:

Prestasi akademik akan dinilai berdasarkan angkatan, mata kuliah, dan nilai rata-rata. Dari data dalam tabel, dapat diambil beberapa kesimpulan berikut ini.

Pada mata kuliah CS01, mahasiswa angkatan 2019 memiliki nilai rata-rata sebesar 80.50, sedikit lebih tinggi dibandingkan dengan mahasiswa angkatan 2020 yang mencapai 80.33. Meskipun terdapat perbedaan kecil ini, perbedaan tersebut dianggap tidak signifikan karena angkatan 2019 memiliki standar deviasi 17.253019 dan angkatan 2020 meemiliki standar deviasi 17.616280 sehingga memiliki persebaran nilai yang sama.

Mata kuliah KAL01 menunjukkan perbedaan yang lebih mencolok antara angkatan 2019 dan 2020. Mahasiswa angkatan 2019 memiliki nilai rata-rata sebesar 85.75, sementara angkatan 2020 hanya mencapai 70.57. Selain itu, angkatan 2019 juga memiliki standar deviasi yang lebih kecil, menandakan bahwa variasi nilai-nilainya lebih dekat dengan rata-rata.

Di sisi lain, pada mata kuliah STA02, mahasiswa angkatan 2019 menunjukkan nilai rata-rata sebesar 54.00, lebih rendah dibandingkan dengan angkatan 2020 yang memiliki nilai sebesar 63.67. Standar deviasi untuk angkatan 2020 terlihat lebih tinggi, menunjukkan bahwa variasi nilai-nilai dari angkatan 2020 lebih tersebar daripada angakatan 2019.

**Kesimpulan**

Terdapat perbedaan prestasi akademik yang nampak pada data antar angaktan mahasiswa 2019 dan 2020. Yang dapat dilihat bahwa angaktn 2019 memiliki tingkat kemampuan yang cenderung lebih tinggi dibantingkan angakatan tahun 2020 dengan nilai standar deviasi atau variasi nilai yang hampir mirip.

# **Task 06: Tambahan Informasi**

## **Code 06 #1**

In [ ]:
# Code Soal Untuk No 06 #1

# merge table dosen dan nilai dengan key: NIDN
merged_df = pd.merge(dosen, nilai, on='NIDN')

# membuat table nilai rata-rata masing-masing kolom table merged_df
gender_mean = merged_df[['Jenis_Kelamin', 'nilai' ]]
gender_mean = gender_mean.groupby(merged_df['Jenis_Kelamin']).mean()[['nilai']]
gender_mean = gender_mean.rename({'nilai': 'mean_nilai'}, axis=1)

# membuat table nilai standar deviasi masing-masing kolom table merged_df
gender_std = merged_df.groupby(merged_df['Jenis_Kelamin']).std()[['nilai']]
gender_std = gender_std.rename({'nilai': 'std_nilai'}, axis=1)

# concat table gender_mean dan gender_std
result_gender_dosen = pd.concat([gender_mean, gender_std], axis=1)
result_gender_dosen

mean_nilai  std_nilai
Jenis_Kelamin                       
Pria            79.227273  13.494026
Wanita          60.090909  16.220077

## **Interpretasi 06 #1**


Dari data yang diberikan:

1. Dosen wanita memiliki rata-rata pemberian nilai sebesar 60.090.
2. Sementara dosen pria memiliki rata-rata pemberian nilai sebesar 79.227.
Dari data tersebut, terlihat bahwa dosen wanita cenderung memberikan nilai yang lebih rendah dibandingkan dengan dosen pria.
3. Rata-rata pemberian nilai dari dosen wanita memiliki standar deviasi terendah (16.22), menunjukkan bahwa variasi data akan semakin tersebar.
4. Rata-rata pemberian nilai dari dosen pria memiliki standar deviasi terendah (13.77), menandakan bahwa nilai-nilai tersebut lebih mendekati rata-rata dan variasi datanya tidak terlalu tersebar.


Kesimpulan:

Kesimpulannya, jenis kelamin dosen memiliki pengaruh terhadap pemberian nilai.

## **Code 06 #2**

In [ ]:
# Code Soal Untuk No 06 #2
nilai_extracted = nilai[['id_mata_kuliah','nilai']]

# membuat table nilai rata-rata masing-masing kolom table nilai_extracted
mean_data = nilai_extracted.groupby('id_mata_kuliah')[['nilai']].agg('mean')
mean_data = mean_data.rename({'nilai': 'mean_nilai'}, axis=1)

# membuat table nilai standar deviasi masing-masing kolom table nilai_extracted
std_data = nilai_extracted.groupby(nilai_extracted['id_mata_kuliah']).std()[['nilai']]
std_data = std_data.rename({'nilai': 'std_nilai'}, axis=1)

# concat
combined_data = pd.concat([mean_data,std_data], axis=1)

matkul_df = pd.DataFrame(nilai['id_mata_kuliah'].value_counts())
matkul_df = matkul_df.rename({'id_mata_kuliah': 'frekuensi_mahasiswa_per_kelas'}, axis=1)
matkul_df.reset_index(inplace=True)
matkul_df = matkul_df.rename(columns = {'index':'matkul'})

matakuliah_df = pd.merge(combined_data, matkul_df, left_on='id_mata_kuliah', right_on='matkul')
matakuliah_df = matakuliah_df.set_index('matkul')
matakuliah_df

mean_nilai  std_nilai  frekuensi_mahasiswa_per_kelas
matkul                                                      
CS01     80.428571  15.883504                              7
KAL01    78.666667  12.804389                             15
KAL02    63.500000   8.103497                              4
STA02    58.142857  19.836234                              7

## **Interpretasi 06 #2**

1. Informasi tambahan yang diperoleh dari proses pemrosesan data menunjukkan bahwa jumlah mahasiswa yang mengambil suatu mata kuliah tidak memiliki dampak signifikan terhadap kinerja rata-rata mahasiswa dalam mata kuliah tersebut.
2. Sebagai contoh, mata kuliah KAL01, yang memiliki jumlah mahasiswa terbanyak dalam satu kelas (15 mahasiswa), memiliki nilai rata-rata sebesar 78.666667. Dengan mempertimbangkan deviasi standar, rata-rata tersebut tampaknya menjadi titik tengah data. Namun demikian, nilai rata-rata tersebut lebih rendah dibandingkan dengan CS01 (dengan 7 mahasiswa) dan KAL02 (dengan 4 mahasiswa).

Kesimpulan:

Jumlah mahasiswa tidak signifikan terhadap kinerja akademik.

## **Code 06 #3**

In [ ]:
# Code Soal Untuk No 06 #3
merged_data = pd.merge(dosen, nilai, on='NIDN')
extracted_data = merged_data[['nama', 'id_mata_kuliah', 'nilai']]

# membuat table nilai rata-rata masing-masing kolom table merged data
mean_data = extracted_data.groupby(['nama', 'id_mata_kuliah'])[['nilai']].mean()
mean_data = mean_data.rename({'nilai': 'mean_nilai'}, axis=1)

# membuat table nilai standar deviasi masing-masing kolom table merged data
std_data = extracted_data.groupby(['nama', 'id_mata_kuliah']).std()[['nilai']]
std_data = std_data.rename({'nilai': 'std_nilai'}, axis=1)

dosen_matakuliah = pd.concat([mean_data, std_data], axis=1)
dosen_matakuliah

mean_nilai  std_nilai
nama          id_mata_kuliah                       
Andri Basuki  KAL01            85.750000   7.741355
Annisa Fitria KAL02            63.500000   8.103497
              STA02            58.142857  19.836234
yudi          KAL01            70.571429  13.010984
zulfikar      CS01             80.428571  15.883504

## **Interpretasi 06 #3**

Ditemukan informasi tambahan dalam data, dari hasil pengolahan data di atas, rata-rata pemberian nilai berdasarkan dosen akan dianalisis.

1.   Dua dosen, yaitu Andri Basuki dan Yudi, mengajar mata kuliah KAL01.
2.   Terdapat perbedaan yang signifikan antara nilai yang diberikan oleh kedua dosen tersebut.
3.   Andri Basuki cenderung memberikan nilai yang lebih tinggi pada mata kuliah KAL01 dibandingkan dengan Yudi.

Kesimpulan:

Dapat disimpulkan bahwa terdapat perbedaan dalam cara pemberian nilai antara dosen yang satu dengan yang lain pada mata kuliah yang sama.

## **Code 06 #4**

In [ ]:
# Code Soal Untuk No 06 #4

# merge data
merged_data1 = pd.merge(mahasiswa, nilai, on='nim')
merged_data1 = merged_data1[['NIDN', 'Gender', 'nilai']]

merged_data2 = pd.merge(merged_data1, dosen, on='NIDN')
merged_data2 = merged_data2[['Jenis_Kelamin', 'Gender', 'nilai']]
merged_data2 = merged_data2.rename(columns={'Jenis_Kelamin': 'Gender_Dosen', 'Gender': 'Gender_Mahasiswa'})

# membuat table nilai rata-rata masing-masing kolom table merged_data2
mean_data = merged_data2.groupby(['Gender_Dosen', 'Gender_Mahasiswa'])[['nilai']].mean()
mean_data = mean_data.rename({'nilai': 'mean_nilai'}, axis=1)

# membuat table nilai rata-rata masing-masing kolom table merged_data2
std_data = merged_data2.groupby(['Gender_Dosen', 'Gender_Mahasiswa']).std()[['nilai']]
std_data = std_data.rename({'nilai': 'std_nilai'}, axis=1)

dosen_gender_df = pd.concat([mean_data, std_data], axis=1)
dosen_gender_df

mean_nilai  std_nilai
Gender_Dosen Gender_Mahasiswa                       
Pria         Pria               83.666667  10.368951
             Wanita             73.900000  15.351077
Wanita       Pria               54.333333   8.640988
             Wanita             67.000000  21.330729

## **Interpretasi 06 #4**

1. Dosen pria cenderung memberikan nilai tinggi kepada mahasiswa (pria), begitu juga dosen wanita kepada mahasiswi (wanita).
2. Namun, variabel lain memengaruhi kriteria penilaian dosen.

Kesimpulan:

Gender dosen berpengaruh terhadap penilaian, tetapi ada faktor lain.

## **Code 06 #5**

In [ ]:
# Code Soal Untuk No 06 #5
mahasiswa_data = mahasiswa[['nim','Asal_Daerah']]
nilai_data = nilai[['nim','nilai']]
merged_data = pd.merge(mahasiswa_data, nilai_data, on='nim')

# membuat table nilai rata-rata masing-masing kolom table merged_data
mean_data = merged_data.groupby('Asal_Daerah')[['nilai']].agg('mean')
mean_data = mean_data.rename({'nilai': 'mean_nilai'}, axis=1)

# membuat table nilai standar deviasi masing-masing kolom table merged_data
std_data = merged_data.groupby(merged_data['Asal_Daerah']).std()[['nilai']]
std_data = std_data.rename({'nilai': 'std_nilai'}, axis=1)

# concat
result_data = pd.concat([mean_data, std_data], axis=1)
result_data

mean_nilai  std_nilai
Asal_Daerah                       
Lokal         75.764706  17.953166
Luar Daerah   69.750000  15.673757

## **Interpretasi 06 #5**

1. Dari dataset tersebut, ditemukan bahwa mahasiswa yang berasal dari daerah lokal memiliki nilai rata-rata yang lebih tinggi dibandingkan dengan mahasiswa yang berasal dari luar daerah.
2. Penekanan pada nilai rata-rata di sini digunakan karena variasi data cenderung kecil, menandakan bahwa nilai rata-rata menjadi representasi yang tepat untuk tengah distribusi data tersebut.

Kesimpulan:

Mahasiswa lokal memiliki rata nilai lebih tinggi daripada mahasiswa non-lokal (luar daerah).